### PR TITLE
Introduce a `Slot` type

### PR DESCRIPTION
**Pull request by bkchr**
Instead of having some type definition that only was used in half of the
code or directly using `u64`, this adds a new unit type wrapper `Slot`.
This makes it especially easy for the outside api to know what type is
expected/returned.

polkadot companion: https://github.com/paritytech/polkadot/pull/2345

### CODE REVIEW

File: test-utils/runtime/src/lib.rs [hunk 1]
```diff
@@ -761,7 +761,7 @@ cfg_if! {
 				}
 
 				fn generate_key_ownership_proof(
-					_slot_number: sp_consensus_babe::SlotNumber,
+					_slot_number: sp_consensus_babe::Slot,
```
Comment (andresilva) [inline test-utils/runtime/src/lib.rs:764]:
```suggestion
					_slot: sp_consensus_babe::Slot,
```

File: test-utils/runtime/src/lib.rs [hunk 2]
```diff
@@ -1020,7 +1020,7 @@ cfg_if! {
...
```
Comment (andresilva) [inline test-utils/runtime/src/lib.rs:1023]:
```suggestion
					_slot: sp_consensus_babe::Slot,
```

File: primitives/consensus/babe/src/lib.rs [hunk 3]
```diff
@@ -356,9 +355,9 @@ pub struct Epoch {
 	/// The epoch index.
 	pub epoch_index: u64,
 	/// The starting slot of the epoch.
-	pub start_slot: SlotNumber,
+	pub start_slot: Slot,
 	/// The duration of this epoch.
-	pub duration: SlotNumber,
+	pub duration: Slot,
```
Comment (andresilva) [inline primitives/consensus/babe/src/lib.rs:360]:
```suggestion
	pub duration: u64,
```

We define `epoch_length` above as `u64` as well.

File: primitives/consensus/babe/src/lib.rs [hunk 4]
```diff
@@ -93,11 +92,11 @@ pub type BabeBlockWeight = u32;
 /// Make a VRF transcript from given randomness, slot number and epoch.
 pub fn make_transcript(
 	randomness: &Randomness,
-	slot_number: u64,
+	slot_number: Slot,
```
Comment (andresilva) [inline primitives/consensus/babe/src/lib.rs:95]:
```suggestion
	slot: Slot,
```

File: primitives/consensus/babe/src/lib.rs [hunk 5]
```diff
@@ -93,11 +92,11 @@ pub type BabeBlockWeight = u32;
...
 	epoch: u64,
 ) -> Transcript {
 	let mut transcript = Transcript::new(&BABE_ENGINE_ID);
-	transcript.append_u64(b"slot number", slot_number);
+	transcript.append_u64(b"slot number", *slot_number);
```
Comment (andresilva) [inline primitives/consensus/babe/src/lib.rs:99]:
```suggestion
	transcript.append_u64(b"slot number", *slot);
```

File: primitives/consensus/babe/src/lib.rs [hunk 6]
```diff
@@ -107,13 +106,13 @@ pub fn make_transcript(
 #[cfg(feature = "std")]
 pub fn make_transcript_data(
 	randomness: &Randomness,
-	slot_number: u64,
+	slot_number: Slot,
```
Comment (andresilva) [inline primitives/consensus/babe/src/lib.rs:109]:
```suggestion
	slot: Slot,
```

File: primitives/consensus/babe/src/lib.rs [hunk 7]
```diff
@@ -107,13 +106,13 @@ pub fn make_transcript(
...
 	epoch: u64,
 ) -> VRFTranscriptData {
 	VRFTranscriptData {
 		label: &BABE_ENGINE_ID,
 		items: vec![
-			("slot number", VRFTranscriptValue::U64(slot_number)),
+			("slot number", VRFTranscriptValue::U64(*slot_number)),
```
Comment (andresilva) [inline primitives/consensus/babe/src/lib.rs:115]:
```suggestion
			("slot number", VRFTranscriptValue::U64(*slot)),
```

File: client/consensus/aura/src/digests.rs [hunk 8]
```diff
@@ -57,11 +58,11 @@ impl<P, Hash> CompatibleDigestItem<P> for DigestItem<Hash> where
 		self.try_to(OpaqueDigestItemId::Seal(&AURA_ENGINE_ID))
 	}
 
-	fn aura_pre_digest(slot_num: u64) -> Self {
+	fn aura_pre_digest(slot_num: Slot) -> Self {
```
Comment (andresilva) [inline client/consensus/aura/src/digests.rs:61]:
```suggestion
	fn aura_pre_digest(slot: Slot) -> Self {
```

File: client/consensus/aura/src/digests.rs [hunk 9]
```diff
@@ -57,11 +58,11 @@ impl<P, Hash> CompatibleDigestItem<P> for DigestItem<Hash> where
...
 		DigestItem::PreRuntime(AURA_ENGINE_ID, slot_num.encode())
```
Comment (andresilva) [inline client/consensus/aura/src/digests.rs:62]:
```suggestion
		DigestItem::PreRuntime(AURA_ENGINE_ID, slot.encode())
```

File: client/consensus/aura/src/digests.rs [hunk 10]
```diff
@@ -38,10 +39,10 @@ pub trait CompatibleDigestItem<P: Pair>: Sized {
 	fn as_aura_seal(&self) -> Option<Signature<P>>;
 
 	/// Construct a digest item which contains the slot number
-	fn aura_pre_digest(slot_num: u64) -> Self;
+	fn aura_pre_digest(slot_num: Slot) -> Self;
```
Comment (andresilva) [inline client/consensus/aura/src/digests.rs:42]:
```suggestion
	fn aura_pre_digest(slot: Slot) -> Self;
```

File: client/consensus/babe/src/tests.rs [hunk 11]
```diff
@@ -572,23 +566,23 @@ fn can_author_block() {
...
+	slot_number: Option<Slot>,
 	proposer_factory: &mut DummyFactory,
 	block_import: &mut BoxBlockImport<TestBlock, Transaction>,
 ) -> sp_core::H256 {
 	let mut proposer = futures::executor::block_on(proposer_factory.init(parent)).unwrap();
 
 	let slot_number = slot_number.unwrap_or_else(|| {
 		let parent_pre_digest = find_pre_digest::<TestBlock>(parent).unwrap();
-		parent_pre_digest.slot_number() + 1
+		Slot(parent_pre_digest.slot().0 + 1)
```
Comment (andresilva) [inline client/consensus/babe/src/tests.rs:577]:
```suggestion
		parent_pre_digest.slot() + 1.into()
```

**Review by andresilva — CHANGES_REQUESTED** [review]
I started suggesting all of the naming changes related to `slot_number` -> `slot`. But this is cumbersome to do on github, do you mind if I just them locally and push a commit for this?

File: client/consensus/babe/src/tests.rs (hunk 11 above)
Comment (bkchr) [reply client/consensus/babe/src/tests.rs:577]: Ahh I added the `Add` impl later :see_no_evil:

File: primitives/consensus/slots/src/lib.rs [hunk 12]
```diff
@@ -21,8 +21,38 @@
 
 use codec::{Decode, Encode};
 
-/// A slot number.
-pub type SlotNumber = u64;
+/// Unit type wrapper that represents a slot.
+#[derive(Debug, Encode, Decode, PartialEq, Eq, PartialOrd, Clone, Copy, Default, Ord)]
+pub struct Slot(pub u64);
```
Comment (andresilva) [inline primitives/consensus/slots/src/lib.rs:26]: What do you think about not exposing the inner value and just using `Deref` instead? For some reason I find code like `slot.0` really ugly 🙈

**Review by andresilva — APPROVED** [review]
Change LGTM, I pushed a commit with some extra renaming. Will give another extra look before merging to double check tiny changes.

**Review by octol — APPROVED** [review]
Looking good! I guess a natural next step could be to also create a type for the slot durations?

**Review by andresilva — APPROVED** [review]
Thanks!

**Comment by bkchr** [comment]
bot merge

**Pull request merged**